### DISCUSSION TITLE
Single UDP Packet

### DISCUSSION THREAD

**Discussion by clownfishnemoGitHub** [dannagle/PacketSender#280] [Q&A]
Hello dannagle

I am new to packetsender and very nice its gui awesome, (only the horizontal slider is to narrow a bit)

how do i perform send a single udp packet ?
i tried to use this as ascii string:
`M-SEARCH * HTTP/1.1\r\nHOST: 239.255.255.250:1900\r\nMAN: ssdp:discover\r\nMX: 10\r\nST: ssdp:all`
To Address: 239.255.255.250
To Port: 1900
Resend Delay: 0
UDP

i tried also to left out the first part of the ascii string:
`MAN: ssdp:discover\r\nMX: 10\r\nST: ssdp:all`

but both did not work out, what is the syntax for the ascii field in case UDP mode is selected by the combo box
is it an idea to disable the fields: To Address and To Port (add "To" infront of the labels of the input textboxes ;)

this ascii works for python, where i use following steps, is it an idea when UDP mode is selected
to enable options to configure/ adjust response buffer size, and an option to set a timeout for waiting for dataSocket receive

```
dataSocket=socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

dataSocket.settimeout(5)

dataSocketRequestSSDP.="M-SEARCH * HTTP/1.1"+"\r\n"+"HOST: 239.255.255.250:1900"+"\r\n"+"MAN: ssdp:discover"+"\r\n"+"MX: 10\r\nST: ssdp:all"
dataUTF8=dataSocketRequestSSDP.encode('utf-8')
dataSocket.sendto(dataUTF8, 0, (239.255.255.250,1900))

#dataLenghtMAX=512
#dataLenghtMAX=1024
#dataLenghtMAX=2048
#dataLenghtMAX=4096
dataLenghtMAX=8192
#dataLenghtMAX=16384
try:
  resultSocketResponse=dataSocket.recv(dataLenghtMAX, 0)
except:
  print('\r')
  print(str(sys.exc_info()[0])+'\r')
sys.exit(0)
```

**Comment by dannagle** [comment]
Hello,

You can convert a python string into hex for easy copy-paste to the "HEX" line of Packet Sender. I looked at what you pasted, and modified it like this...

```python
dataSocketRequestSSDP ="M-SEARCH * HTTP/1.1"+"\r\n"+"HOST: 239.255.255.250:1900"+"\r\n"+"MAN: ssdp:discover"+"\r\n"+"MX: 10\r\nST: ssdp:all"
dataUTF8=dataSocketRequestSSDP.encode('utf-8').hex()
print(dataUTF8)
exit(0)
```
This outputted:
```
% python3 testit.py
4d2d534541524348202a20485454502f312e310d0a484f53543a203233392e3235352e3235352e3235303a313930300d0a4d414e3a20737364703a646973636f7665720d0a4d583a2031300d0a53543a20737364703a616c6c
```


Copy that into the HEX line of Packet Sender to get an exact match of what you are trying to send.

That became this...
```
M-SEARCH * HTTP/1.1\r\nHOST: 239.255.255.250:1900\r\nMAN: ssdp:discover\r\nMX: 10\r\nST: ssdp:all
```

I am guessing you then want to send to address 239.255.255.250 on port 1900.

Make sure you are in IPv4 mode and the UDP service is turned on...

![image](https://user-images.githubusercontent.com/1140573/150197527-3645e3af-4820-4ca9-ae90-5fb248656f99.png)


Note that you do not have to actually join the multicast group to send. You will if you want to receive though...

![image](https://user-images.githubusercontent.com/1140573/150197809-02ac2f52-b42a-49e0-b3b2-9fdd9520c73f.png)

**Comment by clownfishnemoGitHub** [comment]
![afbeelding](https://user-images.githubusercontent.com/96440536/150536450-dd6f21b9-a630-422f-b705-459d08569baa.png)

![afbeelding](https://user-images.githubusercontent.com/96440536/150536672-e095963b-7555-46c7-8a5e-f3b568b1e988.png)
i did not know these buttons, for some reason my screen is unable to display them nicely due almost shady
but left one UDP is turned on and it shows a (port) number
the two middle TCP and SSL i disabled
the right one shows IPv4 Mode

thanks for your hex string and its ascii represenation, thanks to let know the .hex trick for python :)
therefore i know the input fields are correct to perform a M-SEARCH request by packetsender :sunglasses: 
i skipped for now to join the option multicast group i founds its menu also at top menu bar

unfortunenately none answer received when perform Send, yeah send to address 239.255.255.250 on port 1900 thats right
it works for python on my other machine, this machine is a vmware running linux/kali

i used angry ip scanner and it finds the ip of my IOT device connected at my local network at my router
this IOT device when it receives such SSDP multicast message, then this IOT device will report back with an answer message

first i should rules things out further, i go try the python snippet at this machine which vmware running linux/kali
then i hope it fails :), next i should tweak the network options of vmware, but i am very very new to network stuff :sweat_smile:

**Comment by clownfishnemoGitHub** [comment]
![afbeelding](https://user-images.githubusercontent.com/96440536/150562097-bbcf6463-268e-4a9f-95d3-306526970799.png)

hmm my windows laptop and or its vmware linux/kali running are both unable to perform the python _SSDP_multicast.py
an other windows machine also attached to same router is able to run this python script, and it nicely detects the IOT device
by the incoming receiving message/ contents

i think i should look within my network router to check its firewall settings or like wise

**Reply by clownfishnemoGitHub** [reply]
almost there, after fiddling around with my router settings, and router restarts,
and restarted my network adapter at my windows laptop, this make i am now able to ping my IOT device from the windows cmd line

good old ping :)

i started my vmware linux/kali at my windows laptop,
lauched a terminal to perform a ping by hping3 tool to the IOT device it worked, nice one step further

is it possible to use packetsender to create/ perform a ping :) ?
i did a quick search and found this description <https://inc0x0.com/icmp-ip-packets-ping-manually-create-and-send-icmp-ip-packets/>

but at linux/kali my tiny python _SSDP_multicast.py did not work still the socket.timeout this means not received any incoming answer

hmm i searched at the web: "does linux block multicast ipv4"
and it seems it can do.. i am very new to linux
(that could explain why python/ packetsender unable to perfom multicast at linux/kali)
hmm now i think somewhere in my linux/kali there is a configuration item/ setting which blocks multicast
https://duckduckgo.com/?q=does+linux+block+multicast+ipv4&atb=v1-1&ia=web

i go searching and let know

**Reply by clownfishnemoGitHub** [reply]
not to s p a m

https://serverfault.com/questions/1036355/why-224-0-0-1-traffic-is-filtered-by-linux-hosts
https://www.linuxquestions.org/questions/linux-networking-3/how-do-i-disable-multicast-4175593047/
https://www.linuxquestions.org/questions/linux-networking-3/multicast-packets-being-dropped-blocked-4175421702/

this link looks to be very usefull :) but i am such a noob for linux total new lol
https://unix.stackexchange.com/questions/25872/how-can-i-know-if-ip-multicast-is-enabled

i did not know there are different multicast address ranges this one startswith 239, nice to learn about,
multicast address ipv4: 239.255.255.250 port 1900

**Comment by clownfishnemoGitHub** [comment]
```
┌──(root💀kali)-[/media/root/4E74E98F74E97A5B]
└─# ip maddr show      
1:      lo
        inet  224.0.0.1
        inet6 ff02::1
        inet6 ff01::1
2:      eth0
        link  01:00:5e:00:00:01
        link  33:33:00:00:00:01
        link  33:33:ff:2d:94:d4
        link  33:33:ff:96:a2:c6
        inet  224.0.0.1
        inet6 ff02::1:ff96:a2c6
        inet6 ff02::1:ff2d:94d4 users 2
        inet6 ff02::1
        inet6 ff01::1
                                                                         
┌──(root💀kali)-[/media/root/4E74E98F74E97A5B]
└─# netstat -g
IPv6/IPv4 Group Memberships
Interface       RefCnt Group
--------------- ------ ---------------------
lo              1      all-systems.mcast.net
eth0            1      all-systems.mcast.net
lo              1      ip6-allnodes
lo              1      ff01::1
eth0            1      ff02::1:ff96:a2c6
eth0            2      ff02::1:ff2d:94d4
eth0            1      ip6-allnodes
eth0            1      ff01::1
                                                                         
┌──(root💀kali)-[/media/root/4E74E98F74E97A5B]
└─# ip link show eth0 | grep MULTICAST
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
                                                                         
┌──(root💀kali)-[/media/root/4E74E98F74E97A5B]
└─# ifconfig eth0
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet ...  netmask 255.255.255.0  broadcast ...  # i stripped out the ip and related and replaced by ...
        inet6 ...  prefixlen 64  scopeid 0x0<global>
        inet6 ...  prefixlen 64  scopeid 0x20<link>
        inet6 ...   prefixlen 64  scopeid 0x0<global>
        ether ...  txqueuelen 1000  (Ethernet)
        RX packets 790  bytes 82281 (80.3 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 451  bytes 34826 (34.0 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

by the looks it support multicast, hmm then my vmware probarly blocks it i cpuld an other vmware os and see if it works there
using same network vmware settings

**Reply by dannagle** [reply]
At this point, I would try live-booting Linux on bare metal. I'd also disable all network cards except the one to be tested. I've seen packets attempt to go out virtual adapters.

**Reply by clownfishnemoGitHub** [reply]
![afbeelding](https://user-images.githubusercontent.com/96440536/151071974-620c9930-eae5-42ff-90aa-cb2d1a0188a1.png)

![afbeelding](https://user-images.githubusercontent.com/96440536/151073079-981167c1-031f-4072-b590-2ea436e96fa5.png)
nice it works within vmware linux/kali

it turns out i have a second addtional firewall by my v i r u s scanner which blocks udp and ssdp traffic
i turned of my scanner and suddendly all worked, i think i turned it also off when i did initial testing at my windows host machine
by run the python script to check if it first works without vmware linux/kali

the default windows firewall is still activated this introduce none trouble what so ever, still need to check
what this default windows firewall does which rules it uses

i learned a lot thanks for the help,
now i need to re configure my additional firewall provided by the v i r u s scanner, that can be trick due all its complicated options

thanks
